src/libraries/System.Security.Cryptography.ProtectedData/src/System/Security/Cryptography/ProtectedData.cs

```diff
@@ -81,9 +376,26 @@ private static byte[] ProtectOrUnprotect(byte[] inputData, byte[]? optionalEntro
                             throw new OutOfMemoryException();
 
                         int length = (int)(outputBlob.cbData);
-                        byte[] outputBytes = new byte[length];
-                        Marshal.Copy(outputBlob.pbData, outputBytes, 0, length);
-                        return outputBytes;
+                        if (allocateArray)
+                        {
+                            byte[] outputBytes = new byte[length];
+                            Marshal.Copy(outputBlob.pbData, outputBytes, 0, length);
+                            outputData = outputBytes;
+                            bytesWritten = length;
+                            return true;
+                        }
+
+                        if (outputBlob.cbData > outputSpan.Length)
+                        {
+                            bytesWritten = 0;
+                            outputData = null;
+                            return false;
+                        }
+
+                        new Span<byte>(outputBlob.pbData.ToPointer(), length).CopyTo(outputSpan);
+                        bytesWritten = length;
+                        outputData = null;
+                        return true;
                     }
                     finally
                     {
```

Comment:
While we're here, let's CryptographicOperations.ZeroMemory outputBlob before we FreeHGlobal it.  At least, if we're doing Unprotect.

Reply:
Oh, we do, just inefficiently.

Eh, bonus points for replacing the loop with the call to ZeroMemory, but since it's functionally correct I don't super-care.